File: NET/JNetCall.Sharp.Common/Server/AbstractHost.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Threading.Tasks;
using JNetCall.Sharp.API;
using JNetProto.Sharp.Beans;
using JNetProto.Sharp.Tools;

namespace JNetCall.Sharp.Server
{
    public abstract class AbstractHost<T> : IDisposable
    {
        private readonly Type _serviceClass;
        private readonly IDictionary<string, Type> _interfaces;
        protected readonly ProtoSettings _config;

        protected AbstractHost(Type serviceClass)
        {
            _serviceClass = serviceClass;
            _interfaces = new Dictionary<string, Type>();
            _config = new ProtoSettings();
        }

        protected T CreateInst()
        {
            var svc = (T)Activator.CreateInstance(_serviceClass);
            return svc;
        }

        public void AddServiceEndpoint<TSvc>()
        {
            AddServiceEndpoint(typeof(TSvc));
        }

        private void AddServiceEndpoint(Type interfaceClass)
        {
            var name = interfaceClass.Name;
            _interfaces.Add(name, interfaceClass);
        }

        protected void HandleCall(object inst, MethodInfo[] methods, 
            MethodCall call, ProtoConvert proto)
        {
            var callId = call.I;
            if (!_interfaces.ContainsKey(call.C))
            {
                var debug = call.C;
                Write(proto, debug, MethodStatus.ClassNotFound, callId);
                return;
            }
            var callName = call.M;
            var method = methods.FirstOrDefault(m => CheckMethod(m, callName));
            if (method == null)
            {
                var debug = call.C + "::" + call.M;
                Write(proto, debug, MethodStatus.MethodNotFound, callId);
                return;
            }
            try
            {
                var types = method.GetParameters()
                    .Select(p => p.ParameterType).ToArray();
                var args = Conversions.Convert(types, call.A);
                var res = method.Invoke(inst, args);
                if (res is Task task)
                {
                    // TODO Handle non-sync!
                    res = GetTaskResult(task);
                }
                Write(proto, res, MethodStatus.Ok, callId);
            }
            catch (Exception e)
            {
                var cause = e is TargetInvocationException ti
                    ? ti.InnerException
                    : e;
                var debug = cause!.ToString();
                Write(proto, debug, MethodStatus.MethodFailed, callId);
            }
        }

        private static object GetTaskResult(Task task)
        {
            task.Wait();
            var taskType = task.GetType();
            var taskArg = taskType.GetGenericArguments().FirstOrDefault();
            if (taskArg == null)
                return null;
            var raw = taskType.GetProperty("Result")?.GetValue(task);
            if (raw == null || raw.GetType().Name.Equals("VoidTaskResult"))
                return null;
            return raw;
        }

        private const StringComparison Cmp = StringComparison.InvariantCultureIgnoreCase;

        private static bool CheckMethod(MemberInfo m, string callName)
        {
            var mName = m.Name.Replace("_", string.Empty);
            return mName.Equals(callName, Cmp);
        }

        private static void Write(ProtoConvert proto, object res, MethodStatus status, short id)
        {
            var obj = new MethodResult(id, res, (short)status);
            proto.WriteObject(obj);
            proto.Flush();
        }

        public void Dispose()
        {
            _interfaces.Clear();
        }
    }
}

File: JVM/jnetcall-java-common/src/main/java/jnetcall/java/server/AbstractHost.java
package jnetcall.java.server;

import jnetcall.java.api.MethodCall;
import jnetcall.java.api.MethodResult;
import jnetcall.java.api.MethodStatus;
import jnetproto.java.beans.ProtoConvert;
import jnetproto.java.beans.ProtoSettings;
import jnetbase.java.*;
import jnetproto.java.tools.Conversions;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;

abstract class AbstractHost<T> implements AutoCloseable {
    private final Class<T> serviceClass;
    private final Map<String, Class<?>> interfaces;
    protected final ProtoSettings config;

    protected AbstractHost(Class<T> serviceClass) {
        this.serviceClass = serviceClass;
        this.interfaces = new HashMap<>();
        this.config = new ProtoSettings();
    }

    protected T createInst() throws Exception {
        var clazz = this.serviceClass;
        var svc = clazz.getDeclaredConstructor().newInstance();
        return svc;
    }

    public <I> void addServiceEndpoint(Class<I> interfaceClass) {
        var name = interfaceClass.getSimpleName();
        this.interfaces.put(name, interfaceClass);
    }

    protected void handleCall(Object inst, Method[] methods, MethodCall call, ProtoConvert proto)
            throws Exception {
        var callId = call.I();
        if (!interfaces.containsKey(call.C())) {
            var debug = call.C();
            write(proto, debug, MethodStatus.ClassNotFound, callId);
            return;
        }
        var callName = call.M();
        var method = Arrays.stream(methods)
                .filter(m -> checkMethod(m, callName))
                .findFirst().orElse(null);
        if (method == null) {
            var debug = call.C() + "::" + call.M();
            write(proto, debug, MethodStatus.MethodNotFound, callId);
            return;
        }
        try {
            var args = Conversions.convertFor(call.A(), method);
            var res = method.invoke(inst, args);
            if (res instanceof Future<?> task) {
                // TODO Handle non-sync!
                res = getTaskResult(task);
            }
            write(proto, res, MethodStatus.Ok, callId);
        } catch (Throwable e) {
            var cause = e instanceof InvocationTargetException
                    ? e.getCause() : e;
            var debug = Strings.getStackTrace(cause);
            write(proto, debug, MethodStatus.MethodFailed, callId);
        }
    }

    private static Object getTaskResult(Future<?> task)
            throws ExecutionException, InterruptedException {
        var raw = task.get();
        return raw;
    }

    private boolean checkMethod(Method m, String callName) {
        var cName = callName.replace("_", "");
        return m.getName().equalsIgnoreCase(cName);
    }

    private static void write(ProtoConvert proto, Object res, MethodStatus status, short id)
            throws Exception {
        var obj = new MethodResult(id, res, status.getValue());
        proto.writeObject(obj);
        proto.flush();
    }

    @Override
    public void close() {
        this.interfaces.clear();
    }
}
